File: example-code/compile.py
import json
import sys
import time
import requests  # pip install requests

# NOTE: Provide your API key here
api_key = ""

# NOTE: Give your circuit a name here
CIRCUIT_NAME = "cubic_example"

# NOTE: Provide the path to your .tar.gz file, and circuit type
FILE_NAME = "../gnark/cubic.tar.gz"
CIRCUIT_TYPE = "Gnark"

API_VERSION = "v1"
API_URL = f"https://forge.sindri.app/api/{API_VERSION}/"
api_key_querystring = f"?api_key={api_key}"
headers_json = {"Accept": "application/json"}
headers_multipart = {"Accept": "multipart/form-data"}
headers_urlencode = {
    "Accept": "application/json",
    "Content-Type": "application/x-www-form-urlencoded",
}

# Create new circuit
print("Creating circuit...")
response = requests.post(
    API_URL + "circuit/create" + api_key_querystring,
    headers=headers_json,
    data={
        "circuit_name": CIRCUIT_NAME,
        "circuit_type": CIRCUIT_TYPE
    },
)
assert response.status_code == 201, f"Expected status code 201, received {response.status_code}."
circuit_id = response.json().get("circuit_id")  # Obtain circuit_id

# Load the circuit .tar.gz file
files = {"files": open(FILE_NAME, "rb")}

# Upload the circuit file
response = requests.post(
    API_URL + f"circuit/{circuit_id}/uploadfiles" + api_key_querystring,
    headers=headers_multipart,
    files=files,
)
assert response.status_code == 201, f"Expected status code 201, received {response.status_code}."

# Initiate circuit compilation
response = requests.post(API_URL + f"circuit/{circuit_id}/compile" +
                         api_key_querystring,
                         headers=headers_json)
assert response.status_code == 201, f"Expected status code 201, received {response.status_code}."

# Poll circuit detail unitl it has a status of Ready or Failed
TIMEOUT = 600  # timeout after 10 minutes
for i in range(TIMEOUT):
  response = requests.get(
      API_URL + f"circuit/{circuit_id}/detail" + api_key_querystring,
      headers=headers_json,
      params={"include_verification_key": False},
  )
  assert (response.status_code == 200
          ), f"Expected status code 200, received {response.status_code}."
  status = response.json()["status"]
  if status in ["Ready", "Failed"]:
    print(f"Circuit poll exited after {i} seconds with status: {status}")
    break
  time.sleep(1)
else:
  sys.exit("Circuit compile polling timed out")

# Check for compilation issues
if status == "Failed":
  sys.exit("Circuit compilation failed")
else:
  print("Circuit compilation succeeded!")
  print(json.dumps(response.json(), indent=2))

File: noir/pagerank/compile.mjs
// Use dynamic imports for ESM and node REPL compatibility, not necessary otherwise.
const axios = (await import("axios")).default;
const FormData = (await import("form-data")).default;
const fs = (await import("fs")).default;
const process = (await import("process")).default;
const tar = (await import("tar")).default;

// Make sure to provide your actual API key here.
const SINDRI_API_KEY = process.env.SINDRI_API_KEY || "<your-key-here>";

// Use v1 of the Sindri API.
axios.defaults.baseURL = "https://forge.sindri.app/api/v1";
// Authorize all future requests with an `Authorization` header.
axios.defaults.headers.common["Authorization"] = `Bearer ${SINDRI_API_KEY}`;
// Expect 2xx responses for all requests.
axios.defaults.validateStatus = (status) => status >= 200 && status < 300;

// Create a new circuit.
const createResponse = await axios.post(
  "/circuit/create",
  {
    circuit_name: "pagerank",
    circuit_type: "Noir",
  },
  { validateStatus: (status) => status === 201 },
);
const circuitId = createResponse.data.circuit_id;
console.log("Circuit ID:", circuitId);

// Upload the packaged circuit.
const formData = new FormData();
formData.append(
  "files",
  tar.c({ gzip: true, sync: true }, ["circuits/"]).read(),
  {
    filename: "compress.tar.gz",
  },
);
await axios.post(`/circuit/${circuitId}/uploadfiles`, formData);

// Initiate compilation and poll for completion.
await axios.post(`/circuit/${circuitId}/compile`);
let startTime = Date.now();
let circuitDetailResponse;
while (true) {
  circuitDetailResponse = await axios.get(`/circuit/${circuitId}/detail`, {
    params: { include_verification_key: false },
  });
  const { status } = circuitDetailResponse.data;
  const elapsedSeconds = ((Date.now() - startTime) / 1000).toFixed(1);
  if (status === "Ready") {
    console.log(`Polling succeeded after ${elapsedSeconds} seconds.`);
    break;
  } else if (status === "Failed") {
    throw new Error(
      `Polling failed after ${elapsedSeconds} seconds: ${circuitDetailResponse.data.error}.`,
    );
  } else if (Date.now() - startTime > 30 * 60 * 1000) {
    throw new Error("Timed out after 30 minutes.");
  }
  await new Promise((resolve) => setTimeout(resolve, 1000));
}
console.log("Circuit Detail:");
console.log(circuitDetailResponse.data);
